contracts/utils/structs/MaxHeap.sol

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+/**
+ *  TODO:
+ *  - optimizations
+ *  - changeset
+ *  - add tests
+ *  - add docs
+ *  - add base impl link here
+ *
+ * @dev Library for managing a max heap data structure.
+ *
+ * A max heap is a complete binary tree where each node has a value greater than or equal to its children.
+ * The root node contains the maximum value in the heap.
+ *
+ * This library provides functions to insert, update, and remove elements from the max heap, as well as to
+ * retrieve the maximum element (peek) and check the validity of the heap.
+ *
+ * The max heap has the following properties:
+ *
+ * - Insertion: O(log n)
+ * - Deletion of maximum element: O(log n)
+ * - Retrieval of maximum element (peek): O(1)
+ * - Update of an element: O(log n)
+ *
+ * The max heap is implemented using two mappings:
+ * - `tree`: Maps the position in the heap to the item ID.
+ * - `items`: Maps the item ID to its corresponding `Node` struct, which contains the value and heap index.
+ *
+ * Example usage:
+ *
+ * ```solidity
+ * contract Example {
+ *     using MaxHeap for MaxHeap.MaxHeap;
+ *
+ *     MaxHeap.MaxHeap private heap;
+ *
+ *     function addItem(uint256 itemId, uint256 value) public {
+ *         heap.insert(itemId, value);
+ *     }
+ *
+ *     function removeMax() public returns (uint256, uint256) {
+ *         return heap.pop();
+ *     }
+ *
+ *     function getMax() public view returns (uint256, uint256) {
+ *         return heap.peek();
+ *     }
+ * }
+ * ```
+ */
+library MaxHeap {
+    /**
+     * @dev The position doesn't have a _parent as it's the root.
+     */
+    error InvalidPositionZero();
+
+    struct Node {
+        uint256 value;
+        uint256 heapIndex;
+    }
+
+    struct MaxHeap {
+        mapping(uint256 => uint256) tree;
+        mapping(uint256 => Node) items;
```

Comment:
I'm expecting one of these mapping (maybe both) can be implemented using array. That would result in better data locality when we move to verkle.

Reply:
also, using array with a value smaller than uint256 (for `tree`) would compress the data (shared slots)